bssl-compat/source/SSL_get_curve_name.c

```diff
@@ -1,42 +1,99 @@
-static const struct NamedGroup kNamedGroups[] = {
-    {NID_secp224r1, SSL_CURVE_SECP224R1, "P-224", "secp224r1"},
-    {NID_X9_62_prime256v1, SSL_CURVE_SECP256R1, "P-256", "prime256v1"},
-    {NID_secp384r1, SSL_CURVE_SECP384R1, "P-384", "secp384r1"},
-    {NID_secp521r1, SSL_CURVE_SECP521R1, "P-521", "secp521r1"},
-    {NID_X25519, SSL_CURVE_X25519, "X25519", "x25519"},
- // {NID_CECPQ2, SSL_CURVE_CECPQ2, "CECPQ2", "CECPQ2"},
+static const struct KnownCurveCandidate kCurveCandidates[] = {
+    {ossl_NID_sect163k1, 1, "K-163"},
+    {ossl_NID_sect163r1, 2, ""},
+    {ossl_NID_sect163r2, 3, "B-163"},
+    {ossl_NID_sect193r1, 4, ""},
+    {ossl_NID_sect193r2, 5, ""},
+    {ossl_NID_sect233k1, 6, "K-233"},
+    {ossl_NID_sect233r1, 7, "B-233"},
+    {ossl_NID_sect239k1, 8, ""},
+    {ossl_NID_sect283k1, 9, "K-283"},
+    {ossl_NID_sect283r1, 10, "B-283"},
+    {ossl_NID_sect409k1, 11, "K-409"},
+    {ossl_NID_sect409r1, 12, "B-409"},
+    {ossl_NID_sect571k1, 13, "K-571"},
+    {ossl_NID_sect571r1, 14, "B-571"},
+    {ossl_NID_secp160k1, 15, ""},
+    {ossl_NID_secp160r1, 16, ""},
+    {ossl_NID_secp160r2, 17, ""},
+    {ossl_NID_secp192k1, 18, ""},
+    {ossl_NID_X9_62_prime192v1, 19, "P-192"},
+    {ossl_NID_secp224k1, 20, ""},
+    {ossl_NID_secp224r1, 21, "P-224"},
+    {ossl_NID_secp256k1, 22, ""},
+    {ossl_NID_X9_62_prime256v1, 23, "P-256"},
+    {ossl_NID_secp384r1, 24, "P-384"},
+    {ossl_NID_secp521r1, 25, "P-521"},
+    {ossl_NID_brainpoolP256r1, 26, ""},
+    {ossl_NID_brainpoolP384r1, 27, ""},
+    {ossl_NID_brainpoolP512r1, 28, ""},
+    {ossl_NID_X25519, 29, "X25519"},
+    {ossl_NID_X448, 30, ""},
+    {ossl_NID_ffdhe2048, 256, ""},
+    {ossl_NID_ffdhe3072, 257, ""},
+    {ossl_NID_ffdhe4096, 258, ""},
+    {ossl_NID_ffdhe6144, 259, ""},
+    {ossl_NID_ffdhe8192, 260, ""}
 };
 
 
+size_t SSL_get_all_curve_names(const char **out, size_t max_out) {
+    ossl_SSL_CTX* ctx = ossl_SSL_CTX_new(TLS_client_method());
+    if (!ctx) {
+        bssl_compat_warn("Failed to initialize dummy ssl context");
+        return 0;
+    }
+
+    SSL* ssl = SSL_new(ctx);
+    if (!ssl) {
+        ossl_SSL_CTX_free(ctx);
+        bssl_compat_warn("Failed to initialize dummy ssl object");
+        return 0;
+    }
+
+    size_t num_candidate_elements = sizeof(kCurveCandidates) / sizeof(kCurveCandidates[0]);
+    size_t found_curves_count = 0;
+
+    // Iterate through our hardcoded candidates and attempt to set each one.
+    for (size_t i = 0; i < num_candidate_elements; ++i) {
+        const struct KnownCurveCandidate* candidate = &kCurveCandidates[i]; // Pointer to current candidate
+        int nids[] = {candidate->nid};
+        size_t nids_len = 1;
+
+        if (ossl_SSL_set1_groups(ssl, nids, nids_len)) {
+            // Success: OpenSSL knows this curve and can handle it.
+            // Only add to 'out' if there's space.
+            if (found_curves_count < max_out) {
+                out[found_curves_count] = candidate->name;
```

Comment:
I was thinking, do we want to iterate every time `SSL_get_all_curve_names` is called, or only first time, and then store the `out` content statically?

Reply:
Once should be enough, using a function-local static variable to ensure thread safe initialisation.

Reply:
Updated.